compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -469,7 +469,10 @@ impl<'context> Elaborator<'context> {
                             Type::Error
                         }
                     }
-                    (lhs, rhs) => Type::InfixExpr(Box::new(lhs), op, Box::new(rhs)).canonicalize(),
+                    (lhs, rhs) => {
+                        let infix = Type::InfixExpr(Box::new(lhs), op, Box::new(rhs));
+                        Type::CheckedCast(Box::new(infix.clone()), Box::new(infix)).canonicalize()
```

Comment:
We're canonicalizing the `CheckedCast` for two reasons:
1. To deduplicate it if there are multiple, e.g. with nested `InfixExpr`'s
2. Because `canonicalize` is called in `evaluate_to_field_element`, which is how `CheckedCast` is later checked